e2e/_suites/kubernetes-autodiscover/kubernetes_test.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/google/uuid"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/elastic/e2e-testing/cli/shell"
+)
+
+type kubernetesControl struct {
+	config           string
+	Namespace        string
+	createdNamespace bool
+}
+
+func (c kubernetesControl) WithConfig(config string) kubernetesControl {
+	c.config = config
+	return c
+}
+
+func (c kubernetesControl) WithNamespace(ctx context.Context, namespace string) kubernetesControl {
+	if namespace == "" {
+		namespace = "test-" + uuid.New().String()
+		err := c.createNamespace(ctx, namespace)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to create namespace %s", namespace)
+		}
+		c.createdNamespace = true
+	}
+	c.Namespace = namespace
+	return c
+}
+
+func (c kubernetesControl) createNamespace(ctx context.Context, namespace string) error {
+	if namespace == "" {
+		return nil
+	}
+
+	_, err := c.Run(ctx, "create", "namespace", namespace)
+	if err != nil {
+		return fmt.Errorf("namespace creation failed: %w", err)
+	}
+
+	// Wait for default account to be available, if not it is not possible to
+	// deploy pods in this namespace.
+	timeout := 60 * time.Second
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	for {
+		_, err := c.Run(ctx, "get", "serviceaccount", "default")
+		if err == nil {
+			return nil
+		}
+		select {
+		case <-timeoutCtx.Done():
+			return fmt.Errorf("namespace was created but it is not ready after %s: %w", timeout, err)
+		default:
+		}
+		time.Sleep(time.Second)
```

Comment:
We have functions to waitFor using backoff. See https://github.com/elastic/e2e-testing/search?q=GetExponentialBackOff

Reply:
For loops refactored to use these backoffs: 1cee53297409c5634259727f80ac621144d1d5c9

It looks a bit less go native to me, but the code of the operations look cleaner now, so let's go with it. Thanks for the tip.